### PR TITLE
Remove hash routing

### PR DESCRIPTION
**Pull request by Bjvanminnen**
This ended up turning into a pretty sizable set of changes. Here's what's all involved.

Replace /p/ with /projects/. This is done so that we can require login for some of our /projects/ urls while also handling redirection from /p/ urls.
- /projects/playlab/\* and /projects/artist/\* never require login
- /projects/applab and /projects/applab/channel_id/edit do require login
- /projects/applab/channel_id and /projects/applab/channel_id/view do not require login

Redirect from legacy /p urls
- /p/applab#channel_id gets redirected to /projects/applab/channel_id using pushState (no additional server hit) since's it doesn't require login
- /p/applab#channel_id/edit gets hard redirected to /projects/applab/channel_id/edit since it may require login.
- For IE 9, where pushState is not supported, we'll always hard redirect.

Unfortunately, applab already is broken on IE 9 (https://www.pivotaltracker.com/story/show/98663750) so some of this logic is impossible to test on IE 9. I will try to test the artist/playlab project stuff at least.

As of right now, I still have a few TODOs. Some have been addressed and just need to be removed, others I may still need to address.

### CODE REVIEW

File: dashboard/test/ui/features/projectsLegacy.feature [hunk 1]
```diff
@@ -0,0 +1,51 @@
+Feature: Projects
+
+# TODO - do we need to make sure we have a valid channel
+# TODO - IE9
+
+@stop_after_failure
```
Comment (Bjvanminnen) [inline dashboard/test/ui/features/projectsLegacy.feature]: These tests take an unfortunately long time to run. Each example can wait up to 30 seconds to see the redirect. This becomes particularly painful in cases where something goes wrong. @stop_after_failure makes it so that as soon as we get one failure, we stop running other examples.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 2]
```diff
@@ -388,3 +389,14 @@ dashboard.header.showProjectHeader = function () {
     location.href = '/p'; <%# TODO: Can't call project_list_path because it's not in scope %>
   });
 };
+
+dashboard.header.updateTimestamp = function () {
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: Moved this out of project. Trying to get to a place where project is mostly logic/state, and UI things happen more in header.

File: dashboard/test/ui/features/projectsLegacy.feature [hunk 3]
```diff
@@ -0,0 +1,54 @@
+Feature: Projects
+
+# TODO - this isnt going to work in other environments (where this channel_id
+#        doesn't exist
+# TODO - IE9 doesnt have push state, and will behave differently
+
+@stop_after_failure
```
Comment (Bjvanminnen) [inline dashboard/test/ui/features/projectsLegacy.feature]: These tests take an unfortunately long time to run. Each example can wait up to 30 seconds to see the redirect. This becomes particularly painful in cases where something goes wrong. @stop_after_failure makes it so that as soon as we get one failure, we stop running other examples.
Given that + the TODOs, it might make sense to disable this test for now.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 4]
```diff
@@ -373,18 +373,30 @@ dashboard.header.showProjectHeader = function () {
...
     }
   });
 
   $(document).on('click', '.project_list', function () {
-    location.href = '/p'; <%# TODO: Can't call project_list_path because it's not in scope %>
+    location.href = '/projects'; <%# TODO: Can't call project_list_path because it's not in scope %>
   });
 };
+
+dashboard.header.updateTimestamp = function () {
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: Moved this out of project. Trying to get to a place where project is mostly logic/state, and UI things happen more in header.

File: dashboard/app/helpers/levels_helper.rb [hunk 5]
```diff
@@ -105,10 +105,11 @@ def app_options
     # Provide the channel for templated and applab levels.
     set_channel if @level.project_template_level || @level.game == Game.applab
 
+    callouts = params[:share] ? [] : select_and_remember_callouts(params[:show_callouts])
     # Set videos and callouts.
     view_options(
       autoplay_video: select_and_track_autoplay_video,
-      callouts: select_and_remember_callouts(params[:show_callouts])
+      callouts: callouts
```
Comment (Bjvanminnen) [inline dashboard/app/helpers/levels_helper.rb]: Setting callouts to [] in projects#show didn't seem to be sufficient. Are there ever going to be times where we're sharing but still want to see callouts?

File: dashboard/app/controllers/projects_controller.rb [hunk 6]
```diff
@@ -1,14 +1,24 @@
 class ProjectsController < ApplicationController
-  before_filter :authenticate_user!
+  before_filter :authenticate_user!, except: [:show, :edit, :readonly]
+  before_action :set_level, only: [:show, :edit, :readonly]
```
Comment (Bjvanminnen) [inline dashboard/app/controllers/projects_controller.rb]: this was created using levels_controller.rb as inspriration. I'm not sure if it needs some of that other boiler plate as well? Maybe check_authorization?

File: dashboard/test/ui/features/projects.feature [hunk 7]
```diff
@@ -14,3 +12,51 @@ Scenario: Save Project
...
+  Given I am on "http://studio.code.org/"
+  And I am a teacher
+  And I am on "http://studio.code.org/users/sign_in"
+  And I navigate to the last shared URL
+  Then I append "/edit" to the URL
+  And I get redirected to "/projects/applab/([^\/]*?)/view" via "pushState"
+  And I wait to see "#codeWorkspace"
+  And selector "#codeWorkspace" has class "readonly"
+
+  # TODO - maybe we do a remix and/or create new as well
```
Comment (Bjvanminnen) [inline dashboard/test/ui/features/projects.feature]: I do think there's a lot of value in creating some more projects tests. Getting this committed doesnt necessarily need to gate on this, but this does seem like a space with enough complexity/edge cases that creating more tests would be worthwhile.

File: dashboard/app/helpers/levels_helper.rb (hunk 5 above)
Comment (laurelfan) [inline dashboard/app/helpers/levels_helper.rb]: I added a  `readonly_view_options` method that does this (you'd have to rebase to use this)

File: dashboard/config/routes.rb [hunk 8]
```diff
@@ -82,12 +82,24 @@ def redirect_to_teacher_dashboard
...
+        get '/' + key.to_s, to: 'levels#show', key: value[:name], as: key.to_s
       end
       get '/:template', to: 'projects#template'
     end
   end
 
+  resources :projects, path: '/projects/', only: [:index] do
+    collection do
+      ProjectsController::STANDALONE_PROJECTS.each do |key, value|
+        get '/' + key.to_s, to: 'projects#edit', key: key, as: key.to_s + 'project'
```
Comment (laurelfan) [inline dashboard/config/routes.rb]: hm, if you didn't append the 'project' string here and below, the route helper methods won't have project in there twice (I think, do `rake routes` to check)

File: dashboard/test/ui/features/projects.feature (hunk 7 above)
Comment (laurelfan) [inline dashboard/test/ui/features/projects.feature]: :+1:

File: dashboard/test/ui/features/projectsLegacy.feature [hunk 9]
```diff
@@ -0,0 +1,54 @@
+Feature: Projects
+
+# TODO - this isnt going to work in other environments (where this channel_id
```
Comment (laurelfan) [inline dashboard/test/ui/features/projectsLegacy.feature]: can you create the project in the test?
you can also run any arbitrary dashboard code (in any environment that works with tag @dashboard_db_access) if it's too cumbersome to do that through selenium

File: dashboard/config/routes.rb (hunk 8 above)
Comment (Bjvanminnen) [inline dashboard/config/routes.rb]: I think I originally did this because I also had /p versions of the routes and couldn't have duplicates. You're right that it may makes sense to get rid of the + 'project' portion now.

Comment (laurelfan) [inline dashboard/config/routes.rb]: You can probably remove the :as param from the /p routes if you never generate new routes for them (the as stuff is just to generate the helper methods and is not required)

File: dashboard/app/assets/javascripts/header.js.erb [hunk 10]
```diff
@@ -163,33 +163,33 @@ dashboard.buildHeader = function (stageData, progressData, currentLevelId, userI
...
-    + '  </a>'
-      // TODO + '  <a href="#" onclick="return false;" id="print"><i class="fa fa-print"></i></a>'
-    + '</div>'
-    + '<div id="send-to-phone" class="sharing" style="display: none;">'
-    + '  <label for="phone">Enter a US phone number:</label>'
-    + '  <input type="text" id="phone" name="phone">'
-    + '  <button id="phone-submit" onclick="return false;">Send</button>'
-    + '  <div id="phone-charges">A text message will be sent via <a href="http://twilio.com">Twilio</a>. Charges may apply to the recipient.</div>'
-    + '</div></div>'
+      + '<div id="project-share" class="modal-content">'
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/header.js.erb]: It seems awkward to write all of this inline markup.  Could the markup be moved to an .ejs file for convenience and readability?

Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: I think the eventual plan is for this to move to React.  Right now it's awkwardly copy/pasted from `/apps` until this TODO is resolved:

> // TODO: de-dup with apps code once JS common-core work is done.

I.e. there's no good way to include the .ejs template in `/apps` inside of Dashboard.

Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: Duplicated code: [sharing.html.ejs](https://github.com/code-dot-org/code-dot-org/blob/d41a3c8b45b91888d0897a7853ed077511d044f6/apps/src/templates/sharing.html.ejs) (template) and [feedback.js](https://github.com/code-dot-org/code-dot-org/blob/ec76812829e681c83f51ab49b99e8c972c744809/apps/src/feedback.js#L585) (for send-to-phone and print-window).

**Comment by joshlory** [comment]
I've created two PRs into this branch to fix the primary issues @sfilman found during testing.  Since Brent will be back on Monday I will wait for him to review these fixes and coordinate merging this large change.

https://github.com/code-dot-org/code-dot-org/pull/3162: Use server-side-copy when remixing channel-backed shares
https://github.com/code-dot-org/code-dot-org/pull/3164: Force-redirect /p links to /projects
